File: ports/go/theme.go
package hqtui

// A theme is flat and small on purpose: every token here is one a widget
// actually reaches for. Anything deeper is computed, not configured.

type Theme struct {
	Name string
	// Dark is true for palettes designed against a dark terminal background.
	Dark bool

	Background Color
	// Surface is the panel interior, lifted slightly off the page.
	Surface    Color
	Foreground Color
	Muted      Color

	Primary   Color
	Secondary Color
	Accent    Color

	Success Color
	Warning Color
	Danger  Color
	Info    Color

	Border        Color
	BorderFocused Color
	Title         Color

	Selection     Color
	SelectionText Color
	Cursor        Color

	// Graph holds series colors for multi-line graphs, in draw order.
	Graph []Color
	// Heat is the low-to-high ramp for gauges, meters and heat bars.
	Heat []Color
}

func themeDark() Theme {
	return Theme{
		Name: "dark", Dark: true,
		Background: HexN(0x05070a), Surface: HexN(0x0a0e14),
		Foreground: HexN(0xc6d0db), Muted: HexN(0x5a6b7d),
		Primary: HexN(0x58a6ff), Secondary: HexN(0xbd93f9), Accent: HexN(0x56d4dd),
		Success: HexN(0x5fff87), Warning: HexN(0xffd75f), Danger: HexN(0xff6b6b), Info: HexN(0x56d4dd),
		Border: HexN(0x243040), BorderFocused: HexN(0x56d4dd), Title: HexN(0x7ee2ff),
		Selection: HexN(0x1d3a52), SelectionText: HexN(0xe6f2ff), Cursor: HexN(0x56d4dd),
		Graph: []Color{HexN(0x58a6ff), HexN(0x5fff87), HexN(0xff79c6), HexN(0xffd75f), HexN(0x56d4dd), HexN(0xffa657)},
		Heat:  []Color{HexN(0x5fff87), HexN(0xa8ff60), HexN(0xffd75f), HexN(0xffa657), HexN(0xff6b6b)},
	}
}

// variant is the reference's `variant(DARK, {...})`: start from dark and
// override, so anything a palette does not mention is inherited.
func variant(name string, apply func(*Theme)) Theme {
	t := themeDark()
	t.Name = name
	apply(&t)
	return t
}

func themeDracula() Theme {
	return variant("dracula", func(t *Theme) {
		t.Background, t.Surface = HexN(0x191a21), HexN(0x21222c)
		t.Foreground, t.Muted = HexN(0xf8f8f2), HexN(0x6272a4)
		t.Primary, t.Secondary, t.Accent = HexN(0xbd93f9), HexN(0xff79c6), HexN(0x8be9fd)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0x50fa7b), HexN(0xf1fa8c), HexN(0xff5555), HexN(0x8be9fd)
		t.Border, t.BorderFocused, t.Title = HexN(0x44475a), HexN(0xbd93f9), HexN(0xff79c6)
		t.Selection, t.SelectionText = HexN(0x44475a), HexN(0xf8f8f2)
		t.Graph = []Color{HexN(0xbd93f9), HexN(0x50fa7b), HexN(0xff79c6), HexN(0xf1fa8c), HexN(0x8be9fd), HexN(0xffb86c)}
		t.Heat = []Color{HexN(0x50fa7b), HexN(0xf1fa8c), HexN(0xffb86c), HexN(0xff5555)}
	})
}

func themeNord() Theme {
	return variant("nord", func(t *Theme) {
		t.Background, t.Surface = HexN(0x2e3440), HexN(0x333b4a)
		t.Foreground, t.Muted = HexN(0xe5e9f0), HexN(0x7b88a1)
		t.Primary, t.Secondary, t.Accent = HexN(0x88c0d0), HexN(0xb48ead), HexN(0x8fbcbb)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0xa3be8c), HexN(0xebcb8b), HexN(0xbf616a), HexN(0x81a1c1)
		t.Border, t.BorderFocused, t.Title = HexN(0x434c5e), HexN(0x88c0d0), HexN(0x8fbcbb)
		t.Selection, t.SelectionText = HexN(0x434c5e), HexN(0xeceff4)
		t.Graph = []Color{HexN(0x88c0d0), HexN(0xa3be8c), HexN(0xb48ead), HexN(0xebcb8b), HexN(0x81a1c1), HexN(0xd08770)}
		t.Heat = []Color{HexN(0xa3be8c), HexN(0xebcb8b), HexN(0xd08770), HexN(0xbf616a)}
	})
}

func themeTokyoNight() Theme {
	return variant("tokyo-night", func(t *Theme) {
		t.Background, t.Surface = HexN(0x1a1b26), HexN(0x1f2335)
		t.Foreground, t.Muted = HexN(0xc0caf5), HexN(0x565f89)
		t.Primary, t.Secondary, t.Accent = HexN(0x7aa2f7), HexN(0xbb9af7), HexN(0x7dcfff)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0x9ece6a), HexN(0xe0af68), HexN(0xf7768e), HexN(0x7dcfff)
		t.Border, t.BorderFocused, t.Title = HexN(0x2f3549), HexN(0x7aa2f7), HexN(0x7dcfff)
		t.Selection, t.SelectionText = HexN(0x283457), HexN(0xc0caf5)
		t.Graph = []Color{HexN(0x7aa2f7), HexN(0x9ece6a), HexN(0xbb9af7), HexN(0xe0af68), HexN(0x7dcfff), HexN(0xff9e64)}
		t.Heat = []Color{HexN(0x9ece6a), HexN(0xe0af68), HexN(0xff9e64), HexN(0xf7768e)}
	})
}

func themeGruvbox() Theme {
	return variant("gruvbox", func(t *Theme) {
		t.Background, t.Surface = HexN(0x1d2021), HexN(0x282828)
		t.Foreground, t.Muted = HexN(0xebdbb2), HexN(0x928374)
		t.Primary, t.Secondary, t.Accent = HexN(0x83a598), HexN(0xd3869b), HexN(0x8ec07c)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0xb8bb26), HexN(0xfabd2f), HexN(0xfb4934), HexN(0x83a598)
		t.Border, t.BorderFocused, t.Title = HexN(0x3c3836), HexN(0xfabd2f), HexN(0xfabd2f)
		t.Selection, t.SelectionText = HexN(0x3c3836), HexN(0xfbf1c7)
		t.Graph = []Color{HexN(0x83a598), HexN(0xb8bb26), HexN(0xd3869b), HexN(0xfabd2f), HexN(0x8ec07c), HexN(0xfe8019)}
		t.Heat = []Color{HexN(0xb8bb26), HexN(0xfabd2f), HexN(0xfe8019), HexN(0xfb4934)}
	})
}

func themeMatrix() Theme {
	return variant("matrix", func(t *Theme) {
		t.Background, t.Surface = HexN(0x000000), HexN(0x020a02)
		t.Foreground, t.Muted = HexN(0x9dff9d), HexN(0x2f6b2f)
		t.Primary, t.Secondary, t.Accent = HexN(0x00ff41), HexN(0x00c853), HexN(0x7cff7c)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0x00ff41), HexN(0xd4ff00), HexN(0xff3b30), HexN(0x00e5b0)
		t.Border, t.BorderFocused, t.Title = HexN(0x12401f), HexN(0x00ff41), HexN(0x00ff41)
		t.Selection, t.SelectionText = HexN(0x0d2f14), HexN(0xc9ffc9)
		t.Graph = []Color{HexN(0x00ff41), HexN(0x00c853), HexN(0x7cff7c), HexN(0x00e5b0), HexN(0xd4ff00), HexN(0x2f9e44)}
		t.Heat = []Color{HexN(0x0f7a2e), HexN(0x00c853), HexN(0x00ff41), HexN(0xd4ff00)}
	})
}

func themeMonochrome() Theme {
	return variant("monochrome", func(t *Theme) {
		t.Background, t.Surface = HexN(0x000000), HexN(0x0b0b0b)
		t.Foreground, t.Muted = HexN(0xd0d0d0), HexN(0x6e6e6e)
		t.Primary, t.Secondary, t.Accent = HexN(0xffffff), HexN(0xc0c0c0), HexN(0xe0e0e0)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0xe8e8e8), HexN(0xb8b8b8), HexN(0xffffff), HexN(0xa0a0a0)
		t.Border, t.BorderFocused, t.Title = HexN(0x3a3a3a), HexN(0xd0d0d0), HexN(0xffffff)
		t.Selection, t.SelectionText = HexN(0x303030), HexN(0xffffff)
		t.Graph = []Color{HexN(0xffffff), HexN(0xc8c8c8), HexN(0x909090), HexN(0x686868), HexN(0xb0b0b0), HexN(0x808080)}
		t.Heat = []Color{HexN(0x585858), HexN(0x909090), HexN(0xc8c8c8), HexN(0xffffff)}
	})
}

func themeHighContrast() Theme {
	return variant("high-contrast", func(t *Theme) {
		t.Background, t.Surface = HexN(0x000000), HexN(0x000000)
		t.Foreground, t.Muted = HexN(0xffffff), HexN(0xc0c0c0)
		t.Primary, t.Secondary, t.Accent = HexN(0x00ffff), HexN(0xff00ff), HexN(0xffff00)
		t.Success, t.Warning, t.Danger, t.Info = HexN(0x00ff00), HexN(0xffff00), HexN(0xff0000), HexN(0x00ffff)
		t.Border, t.BorderFocused, t.Title = HexN(0xffffff), HexN(0xffff00), HexN(0xffffff)
		t.Selection, t.SelectionText = HexN(0xffffff), HexN(0x000000)
		t.Graph = []Color{HexN(0x00ffff), HexN(0x00ff00), HexN(0xff00ff), HexN(0xffff00), HexN(0xffffff), HexN(0xff8000)}
		t.Heat = []Color{HexN(0x00ff00), HexN(0xffff00), HexN(0xff8000), HexN(0xff0000)}
	})
}

func themeLight() Theme {
	return Theme{
		Name: "light", Dark: false,
		Background: HexN(0xfbfcfd), Surface: HexN(0xffffff),
		Foreground: HexN(0x1c2530), Muted: HexN(0x6b7a8c),
		Primary: HexN(0x0b62d0), Secondary: HexN(0x7c3aed), Accent: HexN(0x0e7490),
		Success: HexN(0x128a3f), Warning: HexN(0xa86a00), Danger: HexN(0xc62828), Info: HexN(0x0e7490),
		Border: HexN(0xd3dbe4), BorderFocused: HexN(0x0b62d0), Title: HexN(0x0b3d78),
		Selection: HexN(0xd6e6fb), SelectionText: HexN(0x0b2545), Cursor: HexN(0x0b62d0),
		Graph: []Color{HexN(0x0b62d0), HexN(0x128a3f), HexN(0xa3348a), HexN(0xa86a00), HexN(0x0e7490), HexN(0xc2410c)},
		Heat:  []Color{HexN(0x128a3f), HexN(0x7aa300), HexN(0xa86a00), HexN(0xc2410c), HexN(0xc62828)},
	}
}

type namedTheme struct {
	Key   string
	Build func() Theme
}

// Themes lists every built-in, in the reference implementation's order. The key
// is what callers pass to ResolveTheme; both "tokyoNight" and "tokyo-night"
// reach the same palette.
var Themes = []namedTheme{
	{"dark", themeDark},
	{"dracula", themeDracula},
	{"nord", themeNord},
	{"tokyoNight", themeTokyoNight},
	{"gruvbox", themeGruvbox},
	{"matrix", themeMatrix},
	{"monochrome", themeMonochrome},
	{"highContrast", themeHighContrast},
	{"light", themeLight},
}

// ResolveTheme looks a theme up by key or by its own Name. Unknown names fall
// back to dark, because a mistyped theme should not stop an app from starting.
func ResolveTheme(name string) Theme {
	for _, t := range Themes {
		if t.Key == name {
			return t.Build()
		}
	}
	for _, t := range Themes {
		if built := t.Build(); built.Name == name {
			return built
		}
	}
	return themeDark()
}

// Elevate is a slightly lifted or dropped shade of the surface, for zebra rows
// and tracks.
func Elevate(t Theme, amount float64) Color {
	towards := HexN(0x000000)
	if t.Dark {
		towards = HexN(0xffffff)
	}
	return t.Surface.Mix(towards, amount)
}

// HeatColor maps a 0-1 ratio along the theme's heat ramp: green when idle, red
// when hot.
func HeatColor(t Theme, ratio float64) Color { return GradientOf(t.Heat).Sample(ratio) }

// SeriesColor is the nth series color, wrapping around. Negative indices wrap
// from the end.
func SeriesColor(t Theme, index int) Color {
	n := len(t.Graph)
	if n == 0 {
		return DefaultColor
	}
	return t.Graph[((index%n)+n)%n]
}

File: ports/python/hqtui/theme.py
"""A theme is flat and small on purpose.

Every token here is one a widget actually reaches for. Anything deeper is
computed, not configured.
"""

from __future__ import annotations

import dataclasses
from dataclasses import dataclass, field

from .color import Color, Gradient

__all__ = [
    "Theme",
    "THEMES",
    "define_theme",
    "elevate",
    "heat_color",
    "resolve_theme",
    "series_color",
]

_hex = Color.hex


@dataclass(frozen=True, slots=True)
class Theme:
    name: str = "dark"
    dark: bool = True
    """True for palettes designed against a dark terminal background."""

    background: Color = field(default_factory=lambda: _hex(0x05070A))
    surface: Color = field(default_factory=lambda: _hex(0x0A0E14))
    """Panel interiors, lifted slightly off the page."""
    foreground: Color = field(default_factory=lambda: _hex(0xC6D0DB))
    muted: Color = field(default_factory=lambda: _hex(0x5A6B7D))

    primary: Color = field(default_factory=lambda: _hex(0x58A6FF))
    secondary: Color = field(default_factory=lambda: _hex(0xBD93F9))
    accent: Color = field(default_factory=lambda: _hex(0x56D4DD))

    success: Color = field(default_factory=lambda: _hex(0x5FFF87))
    warning: Color = field(default_factory=lambda: _hex(0xFFD75F))
    danger: Color = field(default_factory=lambda: _hex(0xFF6B6B))
    info: Color = field(default_factory=lambda: _hex(0x56D4DD))

    border: Color = field(default_factory=lambda: _hex(0x243040))
    border_focused: Color = field(default_factory=lambda: _hex(0x56D4DD))
    title: Color = field(default_factory=lambda: _hex(0x7EE2FF))

    selection: Color = field(default_factory=lambda: _hex(0x1D3A52))
    selection_text: Color = field(default_factory=lambda: _hex(0xE6F2FF))
    cursor: Color = field(default_factory=lambda: _hex(0x56D4DD))

    graph: tuple[Color, ...] = field(
        default_factory=lambda: (
            _hex(0x58A6FF), _hex(0x5FFF87), _hex(0xFF79C6),
            _hex(0xFFD75F), _hex(0x56D4DD), _hex(0xFFA657),
        )
    )
    """Series colors for multi-line graphs, in draw order."""
    heat: tuple[Color, ...] = field(
        default_factory=lambda: (
            _hex(0x5FFF87), _hex(0xA8FF60), _hex(0xFFD75F),
            _hex(0xFFA657), _hex(0xFF6B6B),
        )
    )
    """Low-to-high ramp for gauges, meters and heat bars."""


def define_theme(base: Theme | None = None, **overrides) -> Theme:
    """Build a theme by overriding a base (dark unless you say otherwise)."""
    start = base if base is not None else Theme()
    if "name" not in overrides:
        overrides["name"] = f"{start.name}-custom"
    return dataclasses.replace(start, **overrides)


def _variant(name: str, **overrides) -> Theme:
    """The reference's ``variant(DARK, {...})``: start from dark and override,
    so anything a palette does not mention is inherited."""
    return dataclasses.replace(Theme(), name=name, **overrides)


DARK = Theme()

DRACULA = _variant(
    "dracula",
    background=_hex(0x191A21), surface=_hex(0x21222C),
    foreground=_hex(0xF8F8F2), muted=_hex(0x6272A4),
    primary=_hex(0xBD93F9), secondary=_hex(0xFF79C6), accent=_hex(0x8BE9FD),
    success=_hex(0x50FA7B), warning=_hex(0xF1FA8C), danger=_hex(0xFF5555), info=_hex(0x8BE9FD),
    border=_hex(0x44475A), border_focused=_hex(0xBD93F9), title=_hex(0xFF79C6),
    selection=_hex(0x44475A), selection_text=_hex(0xF8F8F2),
    graph=(_hex(0xBD93F9), _hex(0x50FA7B), _hex(0xFF79C6), _hex(0xF1FA8C), _hex(0x8BE9FD), _hex(0xFFB86C)),
    heat=(_hex(0x50FA7B), _hex(0xF1FA8C), _hex(0xFFB86C), _hex(0xFF5555)),
)

NORD = _variant(
    "nord",
    background=_hex(0x2E3440), surface=_hex(0x333B4A),
    foreground=_hex(0xE5E9F0), muted=_hex(0x7B88A1),
    primary=_hex(0x88C0D0), secondary=_hex(0xB48EAD), accent=_hex(0x8FBCBB),
    success=_hex(0xA3BE8C), warning=_hex(0xEBCB8B), danger=_hex(0xBF616A), info=_hex(0x81A1C1),
    border=_hex(0x434C5E), border_focused=_hex(0x88C0D0), title=_hex(0x8FBCBB),
    selection=_hex(0x434C5E), selection_text=_hex(0xECEFF4),
    graph=(_hex(0x88C0D0), _hex(0xA3BE8C), _hex(0xB48EAD), _hex(0xEBCB8B), _hex(0x81A1C1), _hex(0xD08770)),
    heat=(_hex(0xA3BE8C), _hex(0xEBCB8B), _hex(0xD08770), _hex(0xBF616A)),
)

TOKYO_NIGHT = _variant(
    "tokyo-night",
    background=_hex(0x1A1B26), surface=_hex(0x1F2335),
    foreground=_hex(0xC0CAF5), muted=_hex(0x565F89),
    primary=_hex(0x7AA2F7), secondary=_hex(0xBB9AF7), accent=_hex(0x7DCFFF),
    success=_hex(0x9ECE6A), warning=_hex(0xE0AF68), danger=_hex(0xF7768E), info=_hex(0x7DCFFF),
    border=_hex(0x2F3549), border_focused=_hex(0x7AA2F7), title=_hex(0x7DCFFF),
    selection=_hex(0x283457), selection_text=_hex(0xC0CAF5),
    graph=(_hex(0x7AA2F7), _hex(0x9ECE6A), _hex(0xBB9AF7), _hex(0xE0AF68), _hex(0x7DCFFF), _hex(0xFF9E64)),
    heat=(_hex(0x9ECE6A), _hex(0xE0AF68), _hex(0xFF9E64), _hex(0xF7768E)),
)

GRUVBOX = _variant(
    "gruvbox",
    background=_hex(0x1D2021), surface=_hex(0x282828),
    foreground=_hex(0xEBDBB2), muted=_hex(0x928374),
    primary=_hex(0x83A598), secondary=_hex(0xD3869B), accent=_hex(0x8EC07C),
    success=_hex(0xB8BB26), warning=_hex(0xFABD2F), danger=_hex(0xFB4934), info=_hex(0x83A598),
    border=_hex(0x3C3836), border_focused=_hex(0xFABD2F), title=_hex(0xFABD2F),
    selection=_hex(0x3C3836), selection_text=_hex(0xFBF1C7),
    graph=(_hex(0x83A598), _hex(0xB8BB26), _hex(0xD3869B), _hex(0xFABD2F), _hex(0x8EC07C), _hex(0xFE8019)),
    heat=(_hex(0xB8BB26), _hex(0xFABD2F), _hex(0xFE8019), _hex(0xFB4934)),
)

MATRIX = _variant(
    "matrix",
    background=_hex(0x000000), surface=_hex(0x020A02),
    foreground=_hex(0x9DFF9D), muted=_hex(0x2F6B2F),
    primary=_hex(0x00FF41), secondary=_hex(0x00C853), accent=_hex(0x7CFF7C),
    success=_hex(0x00FF41), warning=_hex(0xD4FF00), danger=_hex(0xFF3B30), info=_hex(0x00E5B0),
    border=_hex(0x12401F), border_focused=_hex(0x00FF41), title=_hex(0x00FF41),
    selection=_hex(0x0D2F14), selection_text=_hex(0xC9FFC9),
    graph=(_hex(0x00FF41), _hex(0x00C853), _hex(0x7CFF7C), _hex(0x00E5B0), _hex(0xD4FF00), _hex(0x2F9E44)),
    heat=(_hex(0x0F7A2E), _hex(0x00C853), _hex(0x00FF41), _hex(0xD4FF00)),
)

MONOCHROME = _variant(
    "monochrome",
    background=_hex(0x000000), surface=_hex(0x0B0B0B),
    foreground=_hex(0xD0D0D0), muted=_hex(0x6E6E6E),
    primary=_hex(0xFFFFFF), secondary=_hex(0xC0C0C0), accent=_hex(0xE0E0E0),
    success=_hex(0xE8E8E8), warning=_hex(0xB8B8B8), danger=_hex(0xFFFFFF), info=_hex(0xA0A0A0),
    border=_hex(0x3A3A3A), border_focused=_hex(0xD0D0D0), title=_hex(0xFFFFFF),
    selection=_hex(0x303030), selection_text=_hex(0xFFFFFF),
    graph=(_hex(0xFFFFFF), _hex(0xC8C8C8), _hex(0x909090), _hex(0x686868), _hex(0xB0B0B0), _hex(0x808080)),
    heat=(_hex(0x585858), _hex(0x909090), _hex(0xC8C8C8), _hex(0xFFFFFF)),
)

HIGH_CONTRAST = _variant(
    "high-contrast",
    background=_hex(0x000000), surface=_hex(0x000000),
    foreground=_hex(0xFFFFFF), muted=_hex(0xC0C0C0),
    primary=_hex(0x00FFFF), secondary=_hex(0xFF00FF), accent=_hex(0xFFFF00),
    success=_hex(0x00FF00), warning=_hex(0xFFFF00), danger=_hex(0xFF0000), info=_hex(0x00FFFF),
    border=_hex(0xFFFFFF), border_focused=_hex(0xFFFF00), title=_hex(0xFFFFFF),
    selection=_hex(0xFFFFFF), selection_text=_hex(0x000000),
    graph=(_hex(0x00FFFF), _hex(0x00FF00), _hex(0xFF00FF), _hex(0xFFFF00), _hex(0xFFFFFF), _hex(0xFF8000)),
    heat=(_hex(0x00FF00), _hex(0xFFFF00), _hex(0xFF8000), _hex(0xFF0000)),
)

LIGHT = Theme(
    name="light", dark=False,
    background=_hex(0xFBFCFD), surface=_hex(0xFFFFFF),
    foreground=_hex(0x1C2530), muted=_hex(0x6B7A8C),
    primary=_hex(0x0B62D0), secondary=_hex(0x7C3AED), accent=_hex(0x0E7490),
    success=_hex(0x128A3F), warning=_hex(0xA86A00), danger=_hex(0xC62828), info=_hex(0x0E7490),
    border=_hex(0xD3DBE4), border_focused=_hex(0x0B62D0), title=_hex(0x0B3D78),
    selection=_hex(0xD6E6FB), selection_text=_hex(0x0B2545), cursor=_hex(0x0B62D0),
    graph=(_hex(0x0B62D0), _hex(0x128A3F), _hex(0xA3348A), _hex(0xA86A00), _hex(0x0E7490), _hex(0xC2410C)),
    heat=(_hex(0x128A3F), _hex(0x7AA300), _hex(0xA86A00), _hex(0xC2410C), _hex(0xC62828)),
)

THEMES: dict[str, Theme] = {
    "dark": DARK,
    "dracula": DRACULA,
    "nord": NORD,
    "tokyoNight": TOKYO_NIGHT,
    "gruvbox": GRUVBOX,
    "matrix": MATRIX,
    "monochrome": MONOCHROME,
    "highContrast": HIGH_CONTRAST,
    "light": LIGHT,
}
"""Every built-in, keyed as the reference names them. Both ``tokyoNight`` and
``tokyo-night`` reach the same palette through :func:`resolve_theme`."""


def resolve_theme(theme: "Theme | str | None") -> Theme:
    """Look a theme up by key or by its own name.

    Unknown names fall back to dark, because a mistyped theme should not stop an
    app from starting.
    """
    if theme is None:
        return DARK
    if isinstance(theme, Theme):
        return theme
    direct = THEMES.get(theme)
    if direct is not None:
        return direct
    for t in THEMES.values():
        if t.name == theme:
            return t
    return DARK


def elevate(theme: Theme, amount: float = 0.06) -> Color:
    """A slightly lifted or dropped shade of the surface, for zebra rows and
    tracks."""
    return theme.surface.mix(_hex(0xFFFFFF) if theme.dark else _hex(0x000000), amount)


def heat_color(theme: Theme, ratio: float) -> Color:
    """Color a 0-1 ratio along the theme's heat ramp: green when idle, red when
    hot."""
    return Gradient(theme.heat).sample(ratio)


def series_color(theme: Theme, index: int) -> Color:
    """The nth series color, wrapping around. Negative indices wrap from the end."""
    n = len(theme.graph)
    if n == 0:
        from .color import DEFAULT_COLOR

        return DEFAULT_COLOR
    return theme.graph[((index % n) + n) % n]
